### PR TITLE
Do not pyccelize files named after a built-in Python module  

### PR DESCRIPTION
**Pull request by mkaddani**
As discussed in #1402 and #1403, this pull request  intends to raise an error when a file is named after a built-in Python package.

### CODE REVIEW

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: tests/internal/test_internal.py [hunk 1]
```diff
@@ -41,6 +41,11 @@ def test_mpi(f):
 def test_openmp(f, language):
     execute_pyccel(f, accelerators=['openmp'], language=language)
 
+def test_file_named_test():
+    with pytest.raises(ValueError) as exc_info:
+        execute_pyccel(fname="test.py")
+    assert str(exc_info.value) == "files called test can cause problems for some compilers and can't be imported from Python. see #issue: 1403"
```
[the remark this answers is not in the document]
Comment (mkaddani) [reply tests/internal/test_internal.py]: https://github.com/pyccel/pyccel/blob/566fd73499aa94795d628b8b706566c410a02b96/tests/pyccel/test_pyccel.py#L1001-L1004

File: pyccel/codegen/pipeline.py [hunk 2]
```diff
@@ -57,88 +57,60 @@ def execute_pyccel(fname, *,
...
+        Name of the generated module. Default is the same name as the translated file.
+    compiler_export_file : str, optional
+        Name of the JSON file to which compiler information is exported. Default is None.
     """
     if fname.endswith('.pyh'):
         syntax_only = True
         if verbose:
             print("Header file recognised, stopping after syntactic stage")
-
+    if fname == "test.py":
```
[the remark this answers is not in the document]
Comment (mkaddani) [reply pyccel/codegen/pipeline.py]: https://github.com/pyccel/pyccel/blob/566fd73499aa94795d628b8b706566c410a02b96/pyccel/codegen/pipeline.py#L112-L113

**Review by EmilyBourne — APPROVED** [review]
Looks good :+1:

**Comment by yguclu** [comment]
If the problem is that we cannot pyccelize a file which has the same name of a built-in module, why should we only treat the case of `test.py`? Here is the complete list of all built-in packages in Python 3.11: https://docs.python.org/3/py-modindex.html.

**Comment by EmilyBourne** [comment]
I agree. `test` is a particular case which comes up a lot. But all the modules in `python_builtin_libs` are potentially problematic. 
https://github.com/pyccel/pyccel/blob/119d5274a1f7c57cfa0ecbb28dc975e1012cc6f2/pyccel/ast/utilities.py#L94-L97

**Review by EmilyBourne — APPROVED** [review]

File: pyccel/codegen/pipeline.py [hunk 3]
```diff
@@ -57,88 +59,62 @@ def execute_pyccel(fname, *,
...
+    compiler_export_file : str, optional
+        Name of the JSON file to which compiler information is exported. Default is None.
     """
     if fname.endswith('.pyh'):
         syntax_only = True
         if verbose:
             print("Header file recognised, stopping after syntactic stage")
 
+    if Path(fname).stem in python_builtin_libs:
+        raise ValueError(f"File called {os.path.basename(fname)} have same name as python built-in package and can cause problems for some compilers and can't be imported from Python. See #1402")
```
Comment (EmilyBourne) [inline pyccel/codegen/pipeline.py:116]:
```suggestion
        raise ValueError(f"File called {os.path.basename(fname)} has the same name as a python built-in package and can't be imported from Python. See #1402")
```
The compiler problem is irrelevant for file names other than "test". There are also a couple of grammar errors here

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

**Comment by mkaddani** [comment]
since i use random just for testing purposes. should i ignore codacy error:
 ``Standard pseudo-random generators are not suitable for security/cryptographic purposes.``

**Comment by EmilyBourne** [comment]
> since i use random just for testing purposes. should i ignore codacy error? `Standard pseudo-random generators are not suitable for security/cryptographic purposes.`

Yes I think you can go ahead and do that. You can disable the error for the line as described here : https://pylint.pycqa.org/en/v2.10.2/user_guide/message-control.html#block-disables

**Comment by mkaddani** [comment]
1. its seems that disabling pylint wasn't enough  to disable codacy warning.
2. the docstring error is weird, i didn't add any bind_c module!!
https://github.com/pyccel/pyccel/actions/runs/5142867750/attempts/1#summary-13920707023

File: tests/pyccel/test_pyccel.py [hunk 4]
```diff
@@ -1002,6 +1002,6 @@ def test_json():
 #------------------------------------------------------------------------------
 def test_reserved_file_name():
     with pytest.raises(ValueError) as exc_info:
-        libname = str(random.choice(tuple(python_builtin_libs))) + ".py"
+        libname = str(random.choice(tuple(python_builtin_libs))) + ".py" # pylint: disable=no-member
```
Comment (EmilyBourne) [inline tests/pyccel/test_pyccel.py:1005]: The error that you are getting is not a no-member error. Please check codacy for the correct keyword

**Comment by EmilyBourne** [comment]
Please resolve the merge conflicts. Otherwise the errors will probably be quite strange. In particular they tend to point to the wrong line

**Comment by EmilyBourne** [comment]
> the docstring error is weird, i didn't add any bind_c module!! https://github.com/pyccel/pyccel/actions/runs/5142867750/attempts/1#summary-13920707023

The objects that are reported were deleted in the last PR to be merged. As you haven't updated your branch the CI sees these as new objects on your branch

**Review by EmilyBourne — APPROVED** [review]

**Comment by yguclu** [comment]
@mkaddani Can you please update the title and description of this PR?

**Comment by mkaddani** [comment]
> @mkaddani Can you please update the title and description of this PR?

done

File: pyccel/codegen/pipeline.py [hunk 5]
```diff
@@ -57,88 +59,62 @@ def execute_pyccel(fname, *,
                    output_name   = None,
                    compiler_export_file = None):
     """
-    Carries out the main steps required to execute pyccel
+    Run Pyccel on the provided code.
+
+    Carry out the main steps required to execute Pyccel
```
Comment (yguclu) [inline pyccel/codegen/pipeline.py:64]: Missing colon:
```suggestion
    Carry out the main steps required to execute Pyccel:
```

File: pyccel/codegen/pipeline.py [hunk 6]
```diff
@@ -57,88 +59,62 @@ def execute_pyccel(fname, *,
...
+    compiler_export_file : str, optional
+        Name of the JSON file to which compiler information is exported. Default is None.
     """
     if fname.endswith('.pyh'):
         syntax_only = True
         if verbose:
             print("Header file recognised, stopping after syntactic stage")
 
+    if Path(fname).stem in python_builtin_libs:
+        raise ValueError(f"File called {os.path.basename(fname)} has the same name as a python built-in package and can't be imported from Python. See #1402")
```
Comment (yguclu) [inline pyccel/codegen/pipeline.py:116]: Please capitalize Python

**Review by yguclu — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 7]
```diff
@@ -8,6 +8,7 @@ All notable changes to this project will be documented in this file.
 ### Fixed
 
 -   #1302 : Raise error message in case of empty class
+-   #1407 : Raise an error if file name matches a Python package module.
```
[the remark this answers is not in the document]
Comment (mkaddani) [reply CHANGELOG.md]: fixed: 5a347dacd2b7fb237dada0781ef04f7e9be0ad2a

File: pyccel/codegen/pipeline.py (hunk 6 above)
Comment (mkaddani) [reply pyccel/codegen/pipeline.py]: updated: 326459443fc716a50a615b298bd35d97be41b4d6

File: CHANGELOG.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply CHANGELOG.md]: Please do not mark a conversation as resolved if you were not the one to open it...

**Review by yguclu — APPROVED** [review]
Thanks! It looks good

**Pull request merged**